# Chinook Database Querying

## Table of Contents
* [Introduction](#Introduction)
* [Summary](#Summary)
* [Initial Database Review](#Initial-Database-Review)
* [Selecting New Artists](#Selecting-New-Artists)
* [Analyzing Employee Sales Performance](#Analyzing-Employee-Sales-Performance)
* [Albums vs. Individual Tracks](#Albums-vs.-Individual-Tracks)
* [Artists on Most Playslists](#Artists-on-Most-Playlists)
* [How Many Tracks Have Been Purchased vs. Not Purchased](#How-Many-Tracks-Have-Been-Purchased-vs.-Not-Purchased)
* [Conclusion](#Conclusion)

## Introduction

A music distributing company, Chinook, is asking us to analyze their database, chinook.db, to answer business questions. We will be recommending 3 new artists to add to their store, analyzing employee sales performance, determining if the company should continue adding complete albums to the store, uncovering which artists are on the most playlists, and figuring out how many tracks have been purchased and not purchased.

## Summary
* The new artists should be Punk, Rock, and Pop musicians
* Length of employment was the primary determining factor in total sales
* The company should continue selling complete albums
* Mônica Marianno and The 12 Cellists of The Berlin Philharmonic are the two artists on the most playlists
* 48 percent of the tracks in the database have been purchased

## Connecting Database to Notebook

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

## Initial Database Review

Before starting our querying of the database, we will see what tables and views are in the database.

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

Now that we see the names of the tables, we will familiarize ourselves with them by listing their columns and first 3 lines.

In [3]:
%%sql
SELECT * 
  FROM album
 LIMIT 3;

Done.


[(1, 'For Those About To Rock We Salute You', 1),
 (2, 'Balls to the Wall', 2),
 (3, 'Restless and Wild', 2)]

In [4]:
%%sql
SELECT * 
  FROM artist
 LIMIT 3;

Done.


[(1, 'AC/DC'), (2, 'Accept'), (3, 'Aerosmith')]

In [5]:
%%sql
SELECT * 
  FROM customer
 LIMIT 3;

Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3)]

In [6]:
%%sql
SELECT * 
  FROM employee
 LIMIT 3;

Done.


[(1, 'Adams', 'Andrew', 'General Manager', None, '1962-02-18 00:00:00', '2016-08-14 00:00:00', '11120 Jasper Ave NW', 'Edmonton', 'AB', 'Canada', 'T5K 2N1', '+1 (780) 428-9482', '+1 (780) 428-3457', 'andrew@chinookcorp.com'),
 (2, 'Edwards', 'Nancy', 'Sales Manager', 1, '1958-12-08 00:00:00', '2016-05-01 00:00:00', '825 8 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 2T3', '+1 (403) 262-3443', '+1 (403) 262-3322', 'nancy@chinookcorp.com'),
 (3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com')]

In [7]:
%%sql
SELECT * 
  FROM genre
 LIMIT 3;

Done.


[(1, 'Rock'), (2, 'Jazz'), (3, 'Metal')]

In [8]:
%%sql
SELECT * 
  FROM invoice
 LIMIT 3;

Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84),
 (2, 30, '2017-01-03 00:00:00', '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', 9.9),
 (3, 40, '2017-01-05 00:00:00', '8, Rue Hanovre', 'Paris', 'None', 'France', '75002', 1.98)]

In [9]:
%%sql
SELECT * 
  FROM invoice_line
 LIMIT 3;

Done.


[(1, 1, 1158, 0.99, 1), (2, 1, 1159, 0.99, 1), (3, 1, 1160, 0.99, 1)]

In [10]:
%%sql
SELECT * 
  FROM media_type
 LIMIT 3;

Done.


[(1, 'MPEG audio file'),
 (2, 'Protected AAC audio file'),
 (3, 'Protected MPEG-4 video file')]

In [11]:
%%sql
SELECT * 
  FROM playlist
 LIMIT 3;

Done.


[(1, 'Music'), (2, 'Movies'), (3, 'TV Shows')]

In [12]:
%%sql
SELECT * 
  FROM playlist_track
 LIMIT 3;

Done.


[(1, 3402), (1, 3389), (1, 3390)]

In [13]:
%%sql
SELECT * 
  FROM track
 LIMIT 3;

Done.


[(1, 'For Those About To Rock (We Salute You)', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 343719, 11170334, 0.99),
 (2, 'Balls to the Wall', 2, 2, 1, None, 342562, 5510424, 0.99),
 (3, 'Fast As a Shark', 3, 2, 1, 'F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman', 230619, 3990994, 0.99)]

Now that we can see each table in our database, we will begin our analysis.

## Selecting New Aritists

Chinook wants to know which 3 new artists should be added to their store. Chinook will base this decision on which 3 genres sell the best in the USA. We will do this by creating a query that segments the American customers from the database, and grouping the artists they buy from by genre to determine which genre sells the most. The genres in question are Hip-Hop, Punk, Pop, and Blues.

In [14]:
%%sql
WITH 

    usa_tracks AS
    (
        SELECT COUNT(il.track_id) total, i.billing_country country
          FROM invoice_line il
         INNER JOIN invoice i ON i.invoice_id = il.invoice_id
         WHERE i.billing_country = 'USA'
    ),
    
    usa_genres AS
    (
        SELECT 
            i.billing_country AS country,
            g.name AS genre,
            COUNT(*) AS genre_tracks
          FROM customer c
         INNER JOIN invoice i ON i.customer_id = c.customer_id 
         INNER JOIN invoice_line il ON il.invoice_id = i.invoice_id 
         INNER JOIN track t ON t.track_id = il.track_id
         INNER JOIN genre g ON g.genre_id = t.genre_id
         WHERE country = 'USA'
         GROUP BY genre
         ORDER BY genre_tracks DESC
    )
    
SELECT 
    ug.country AS country,
    ug.genre AS genre,
    ug.genre_tracks AS genre_tracks,
    ROUND(CAST(ug.genre_tracks AS Float) / ut.total * 100, 2) AS percent_total_tracks
  FROM usa_genres ug
 INNER JOIN usa_tracks ut ON ut.country = ug.country
 WHERE genre = 'Hip Hop/Rap' OR genre = 'Alternative & Punk' OR genre = 'Pop' OR genre = 'Blues';

Done.


[('USA', 'Alternative & Punk', 130, 12.37),
 ('USA', 'Blues', 36, 3.43),
 ('USA', 'Pop', 22, 2.09),
 ('USA', 'Hip Hop/Rap', 20, 1.9)]

We can see that, of the 4 requested genres, Hip-Hop has the fewest tracks sold in the USA at 20 total tracks and only accounts for 1.9 percent of all track sales. Blues and punk by far sell the most tracks, especially punk, as the `'Alternative & Punk'` track sales reach 130, accounting for 12.37 percent of all sales. Pop comes in third, though by a very narrow margin of 2 total tracks. This only accounts for a .19 percent difference in total tracks sold. This suggests that chosing the last artist by sales per genre could be risky and it may be advantagious to choose based off which artist the Chinook team feels has the most potential. That being said, based off the data presented, we recommend choosing the Pop artist over the Hip-Hop artist.

## Analyzing Employee Sales Performance

We will now query the database to see which employees are generating the most sales, and why some employees may be more successful than others.

In [15]:
%%sql
SELECT
    e.first_name || ' ' || e.last_name AS sales_rep,
    e.birthdate AS birth_date,
    e.hire_date AS hire_date,
    e.city AS work_location,
    ROUND(SUM(i.total), 2) AS total_sales
  FROM invoice i
 INNER JOIN customer c ON c.customer_id = i.customer_id
 INNER JOIN employee e ON e.employee_id = c.support_rep_id
 GROUP BY sales_rep
 ORDER BY total_sales DESC;

Done.


[('Jane Peacock', '1973-08-29 00:00:00', '2017-04-01 00:00:00', 'Calgary', 1731.51),
 ('Margaret Park', '1947-09-19 00:00:00', '2017-05-03 00:00:00', 'Calgary', 1584.0),
 ('Steve Johnson', '1965-03-03 00:00:00', '2017-10-17 00:00:00', 'Calgary', 1393.92)]

We can see that Jane Peacock is the highest performing sales rep at the company. Since she's the youngest of the three, and Margaret Park, the second highest performer, is the oldest, there's no correlation between age and job performance. All 3 employees are based out of Calgary as well, so the results aren't skewed based on location. One thing that is true of their performance, however, is that more recent hires have lower sales numbers. Therefore, we can conclude that longer employee tenure is the primary source of higher sales numbers.

## Sales by Country 

We will now analyze the database to determine the total number of customers, total value of sales per customer, and the average order value per sale in each country in the database. Each country with only one customer will be grouped as `'other'` in our report.

In [16]:
%%sql
WITH 

    country_customers AS 
    (
        SELECT 
            country,
            COUNT(*) AS customers,
            CASE
                WHEN COUNT(*) = 1 THEN 'Other'
                ELSE country
            END AS countries
          FROM customer
         GROUP BY country
    ),
    
    totals AS
    (
        SELECT 
            c.country,
            ROUND(SUM(i.total), 2) total_value,
            COUNT(*) total_sales,
            ROUND(SUM(i.total) / COUNT(*), 2) AS avg_value
        FROM invoice i
        INNER JOIN customer c ON c.customer_id = i.customer_id
        GROUP BY c.country
        ORDER BY total_value DESC
    )

SELECT 
    cc.countries, 
    sum(cc.customers) AS customers, 
    t.total_sales,
    t.total_value, 
    t.avg_value,
    ROUND(CAST(t.total_sales AS Float) / sum(cc.customers), 2) AS avg_customer_sales
  FROM country_customers cc
 INNER JOIN totals t ON t.country = cc.country
 GROUP BY countries
 ORDER BY total_value DESC;

Done.


[('USA', 13, 131, 1040.49, 7.94, 10.08),
 ('Canada', 8, 76, 535.59, 7.05, 9.5),
 ('Brazil', 5, 61, 427.68, 7.01, 12.2),
 ('France', 5, 50, 389.07, 7.78, 10.0),
 ('Germany', 4, 41, 334.62, 8.16, 10.25),
 ('Czech Republic', 2, 30, 273.24, 9.11, 15.0),
 ('United Kingdom', 3, 28, 245.52, 8.77, 9.33),
 ('Portugal', 2, 29, 185.13, 6.38, 14.5),
 ('India', 2, 21, 183.15, 8.72, 10.5),
 ('Other', 15, 10, 75.24, 7.52, 0.67)]

We can see that the country with the most total sales and average sales per customer is `'USA'`, followed by `'Canada'`. The highest average number of customer purchases and value of said purchases is held by `'Czech Republic'`

## Albums vs. Individual Tracks

We will now analyze the database to determine the percentage of purchases that are individual tracks vs. complete albums. This will help Chinook determine if they should continue buying entire albums to maximize their profits or if they should begin purchasing only the most popular tracks off of albums.

In [17]:
%%sql
WITH 
invoices_first_track AS 
(
    SELECT invoice_id, MIN(track_id) AS first_track_id
      FROM invoice_line
  GROUP BY invoice_id
),
invoices_albums AS
(
    SELECT ift.*,
           CASE
           WHEN
                (
                  SELECT il.track_id 
                    FROM invoice_line AS il 
                   WHERE il.invoice_id = ift.invoice_id
                    
                  EXCEPT
                    
                  SELECT track_id 
                    FROM track 
                   WHERE album_id =
                      (
                        SELECT album_id 
                          FROM track 
                         WHERE track_id = ift.first_track_id
                      )
                ) IS NULL
    
                AND 
    
                (
                  SELECT track_id 
                    FROM track 
                   WHERE album_id =
                    (
                        SELECT album_id 
                          FROM track 
                         WHERE track_id = ift.first_track_id
                    )
                    
                 EXCEPT
                    
                 SELECT il.track_id 
                   FROM invoice_line AS il 
                  WHERE il.invoice_id = ift.invoice_id
               ) IS NULL THEN 'YES'
           ELSE 'NO'
           END AS album_purchase
    FROM invoices_first_track AS ift
)
  SELECT
          album_purchase,
          COUNT(invoice_id) AS number_of_invoices,
          ROUND(CAST(COUNT(invoice_id) AS FLOAT) / 
                        (SELECT COUNT(*) FROM invoice),2) * 100 AS percent_of_invoices
    FROM invoices_albums 
GROUP BY album_purchase;

Done.


[('NO', 500, 81.0), ('YES', 114, 19.0)]

The number of invoices containing albums purchased in their entirety is 114, accounting for 19 percent of all purchases. Given almost one in every five purchases is a complete album, it is our official recommendation that Chinook continues to purchase albums in their entirety.

## Artists on Most Playlists

We will now analyze the database to determine which artists show up most frequently on users playlists.

playlist.playlist_id to playlist_track.playlist_id
playlist_track.track to track.track_id
track.artist_id to artist.name

In [18]:
%%sql
SELECT
    a.name artist_name,
    COUNT(*) AS playlist_appearances
  FROM artist a
 INNER JOIN album al ON al.artist_id = a.artist_id 
 INNER JOIN track t ON t.album_id = al.album_id
 INNER JOIN playlist_track pt ON pt.track_id = t.track_id
 INNER JOIN playlist pl ON pl.playlist_id = pt.playlist_id
 GROUP BY pl.playlist_id
 ORDER BY playlist_appearances DESC;

Done.


[('Mônica Marianno', 3290),
 ('The 12 Cellists of The Berlin Philharmonic', 3290),
 ('Philip Glass Ensemble', 1477),
 ('The Office', 213),
 ('The Office', 213),
 ('Aaron Copland & London Symphony Orchestra', 75),
 ('Antônio Carlos Jobim', 39),
 ('Scorpions', 26),
 ('Philip Glass Ensemble', 25),
 ('Berliner Philharmoniker & Herbert Von Karajan', 25),
 ('Aaron Copland & London Symphony Orchestra', 25),
 ('Nirvana', 15),
 ('Audioslave', 1),
 ('Miles Davis', 1)]

We can see that Mônica Marianno and The 12 Cellists of The Berlin Philharmonic appear on the most playlists, with both showing up on 3290 playlists.

## How Many Tracks Have Been Purchased vs. Not Purchased

We will now determine how many tracks in the database have not been purchased. 

In [19]:
%%sql
    WITH not_purchased AS
    (
        SELECT
            track_id
          FROM track

        EXCEPT

        SELECT
            track_id
          FROM invoice_line
    ),
    
    total_tracks AS
    (
        SELECT
            COUNT(*) AS tracks
          FROM track
    )

SELECT 
    COUNT(*) AS total_not_purchased,
    ROUND(CAST(COUNT(*) AS Float) / (
        SELECT
            tracks
          FROM total_tracks
    ), 2) * 100 AS percent_not_purchased
  FROM not_purchased;

Done.


[(1697, 48.0)]

We can see that 1697 total tracks have yet to be purchased, which accounts for 48 percent of the entire database.

## Conclusion

After analyzing the database, we can conclude the following:
* The new artists should be Punk, Rock, and Pop musicians
* Length of employment was the primary determining factor in total sales
* The company should continue selling complete albums
* Mônica Marianno and The 12 Cellists of The Berlin Philharmonic are the two artists on the most playlists
* 48 percent of the tracks in the database have been purchased